// local-validations/local-validations.js
import { registerDecorator } from 'class-validator';

export function AtLeastOneNumber(property, validationOptions) {
  return function a(object, propertyName) {
    registerDecorator({
      name: 'AtLeastOneNumber',
      target: object.constructor,
      propertyName,
      constraints: [property],
      options: validationOptions,
      validator: {
        validate(value, args) {
          const [relatedPropertyName] = args.constraints;
          const relatedValue = (args.object)[relatedPropertyName];
          return typeof value === 'string' &&
            typeof relatedValue === 'string' &&
            value.length > relatedValue.length;
        },
      },
    });
  };
}

export function AtLeastOneLetter(property, validationOptions) {
  return function b(object, propertyName) {
    registerDecorator({
      name: 'AtLeastOneLetter',
      target: object.constructor,
      propertyName,
      constraints: [property],
      options: validationOptions,
      validator: {
        validate(value, args) {
          const [relatedPropertyName] = args.constraints;
          const relatedValue = (args.object)[relatedPropertyName];
          return typeof value === 'string' &&
            typeof relatedValue === 'string' &&
            value.length > relatedValue.length;
        },
      },
    });
  };
}
